### PR TITLE
Add test suite and minor refinements to the utility subsystem

### PR DESCRIPTION
**Pull request by coriolinus**

### CODE REVIEW

File: node/subsystem/src/util.rs [hunk 1]
```diff
@@ -540,17 +540,20 @@ where
...
 					.drain()
 					.map(|(_, handle)| handle.stop())
-					.collect::<FuturesUnordered<_>>();
-				// now wait for all the futures to complete; collect a vector of their results
-				// this is strictly less efficient than draining them into oblivion, but this compiles, and that doesn't
-				// https://github.com/paritytech/polkadot/pull/1376#pullrequestreview-446488645
-				let _ = async move { unordered.collect::<Vec<_>>() }.await;
+					.collect::<FuturesUnordered<_>>()
+					.map(Ok)
+					.forward(drain())
```
Comment (drahnr) [inline node/subsystem/src/util.rs:552]: :+1:

File: node/test-helpers/subsystem/src/lib.rs [hunk 2]
```diff
@@ -15,215 +15,7 @@
...
-			rx: overseer_rx,
-			spawn,
-		},
-		TestSubsystemContextHandle {
-			tx: overseer_tx,
-			rx: all_messages_rx
-		},
-	)
-}
+pub use polkadot_subsystem::test_helpers::*;
```
Comment (ordian) [inline node/test-helpers/subsystem/src/lib.rs:21]: don't think this is going to work given test_helpers is `cfg(test)` gated
substrate uses features approach: https://github.com/paritytech/substrate/search?l=TOML&q=test-helpers
this is prone to accidental feature leaking ofc and needs to be used with care

Comment (coriolinus) [reply node/test-helpers/subsystem/src/lib.rs:21]: To my mind that's a feature, not a bug: it reduces the chance that anybody accidentally uses a test helper in production code.

Comment (ordian) [reply node/test-helpers/subsystem/src/lib.rs:21]: I meant it doesn't compile :)

Comment (coriolinus) [reply node/test-helpers/subsystem/src/lib.rs:21]: Oh, I see. Right, `cfg(test)` doesn't propagate. I'll fix that.

**Comment by coriolinus** [comment]
Interestingly, the same test fails in two runs on CI, but I can't get it to fail locally on Ubuntu or OSX over several runs. I'd prefer not to just delete the test. Does anyone here have experience with failures like this?

File: node/subsystem/src/util.rs [hunk 3]
```diff
@@ -323,8 +331,8 @@ pub trait JobTrait: Unpin {
 	fn run(
 		parent: Hash,
 		run_args: Self::RunArgs,
-		rx_to: mpsc::Receiver<Self::ToJob>,
-		tx_from: mpsc::Sender<Self::FromJob>,
+		receiver: mpsc::Receiver<Self::ToJob>,
+		sender: mpsc::Sender<Self::FromJob>,
```
Comment (drahnr) [inline node/subsystem/src/util.rs:335]: :+1:

File: node/subsystem/src/util.rs [hunk 4]
```diff
@@ -383,12 +420,23 @@ impl<Spawner: Spawn, Job: JobTrait> Jobs<Spawner, Job> {
 					parent_hash,
 					e,
 				);
+
+				if let Some(mut err_tx) = err_tx {
+					// if we can't send the notification of error on the error channel, then
+					// there's no point trying to propagate this error onto the channel too
+					let _ = err_tx.send((Some(parent_hash), JobsError::Job(e))).await;
```
Comment (drahnr) [inline node/subsystem/src/util.rs:427]: should we print a warning here at least?

File: node/subsystem/src/util.rs [hunk 5]
```diff
@@ -356,25 +376,42 @@ pub struct Jobs<Spawner, Job: JobTrait> {
...
 		}
 	}
 
+	/// Monitor errors which may occur during handling of a spawned job.
+	///
+	/// By default, an error in a job is simply logged. Once this is called,
+	/// the error is forwarded onto the provided channel.
+	///
+	/// Errors if the error channel already exists.
+	pub fn fwd_errors(&mut self, tx: mpsc::Sender<(Option<Hash>, JobsError<Job::Error>)>) -> Result<(), Error> {
```
Comment (rphmeier) [inline node/subsystem/src/util.rs:400]: prefer `forward_errors`

File: node/subsystem/src/util.rs [hunk 6]
```diff
@@ -489,9 +538,22 @@ where
...
```
Comment (rphmeier) [inline node/subsystem/src/util.rs:551]: same

File: node/subsystem/src/lib.rs [hunk 7]
```diff
@@ -71,7 +73,7 @@ pub enum FromOverseer<M> {
 ///   * Subsystems dying when they are not expected to
 ///   * Subsystems not dying when they are told to die
 ///   * etc.
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (ordian) [inline node/subsystem/src/lib.rs:76]: I wonder why we derive `PartialEq`, but not `Eq` on some types.
The reason `PartialEq` exists is floating point (NaN != NaN): https://doc.rust-lang.org/std/cmp/trait.PartialEq.html. 

So I think we should also derive `Eq` wherever we derive `Eq`.

Comment (coriolinus) [reply node/subsystem/src/lib.rs]: https://github.com/paritytech/polkadot/pull/1403/commits/5b464b0b4d90e9728faffcfc6d6fcda36d3d97ea

File: node/subsystem/src/messages.rs [hunk 8]
```diff
@@ -408,4 +408,10 @@ pub enum AllMessages {
 	AvailabilityStore(AvailabilityStoreMessage),
 	/// Message for the network bridge subsystem.
 	NetworkBridge(NetworkBridgeMessage),
+	/// Test message
+	///
+	/// This variant is only valid while testing, but makes the process of testing the
+	/// subsystem job manager much simpler.
+	#[cfg(test)]
```
Comment (ordian) [inline node/subsystem/src/messages.rs:415]: do we want `#[cfg(any(test, feature = "test-helpers"))]` here?

Comment (coriolinus) [reply node/subsystem/src/messages.rs:415]: No: this is only used within the internal unit tests of the job manager. If we come up with a use for it later, we can add it then. Until then, I think YAGNI is the better part of valor.

**Pull request merged**